File: DesignPatterns/PrototypePattern/ColorClient.java
import java.util.HashMap; 
import java.util.Map;

abstract class Color implements Cloneable {
  protected String colorName;

  abstract void addColor();

  public Object clone(){
    Object clone = null;
    try {
      clone = super.clone();
    } catch(CloneNotSupportedException e){
      e.printStackTrace();
    }
    return clone;
  }
}

class BlueColor extends Color {

  public BlueColor(){
    this.colorName = "Blue";
  }

  public void addColor(){
    System.out.println("Blue color added");
  }
}

class BlackColor extends Color {
  
  public BlackColor(){
    this.colorName = "Black";
  }

  public void addColor(){
    System.out.println("Black color added");
  }
}

class ColorStore {

  private static Map<String, Color> colorMap = new HashMap<String, Color>();

  static {
    colorMap.put("blue", new BlueColor());
    colorMap.put("black", new BlackColor());
  }

  public static Color getColor(String colorName) {
    return (Color) colorMap.get(colorName).clone();
  }
}

public class ColorClient{
  public static void main(String[] args) {
    System.out.println(ColorStore.getColor("blue"));
    ColorStore.getColor("blue").addColor();

    System.out.println(ColorStore.getColor("black"));
    ColorStore.getColor("black").addColor();

    System.out.println(ColorStore.getColor("blue")); //Verify the new cloned object is returned
    ColorStore.getColor("blue").addColor();

    System.out.println(ColorStore.getColor("black")); //Verify the new cloned object is returned
    ColorStore.getColor("black").addColor();
  }
}

File: DesignPatterns/Creational/Prototype/Ruby/color_client.rb
module Constants
  BLUE = "blue".freeze
  BLACK = "black".freeze
end
class Color
  attr_reader :name

  def initialize(name)
    @name = name
  end

  def add_color
    raise NotImplementedError
  end
end

class BlueColor < Color
  def add_color
    puts "Blue color added"
  end
end

class BlackColor < Color
  def add_color
    puts "Black color added"
  end
end

class ColorStore
  @@color_map = {}
  @@color_map[Constants::BLUE] = BlueColor.new(Constants::BLUE)
  @@color_map[Constants::BLACK] = BlackColor.new(Constants::BLACK)

  def self.fetch_instance(color)
    @@color_map[color].clone()
  end
end

puts ColorStore.fetch_instance(Constants::BLUE)
ColorStore.fetch_instance(Constants::BLUE).add_color

puts ColorStore.fetch_instance(Constants::BLACK)
ColorStore.fetch_instance(Constants::BLACK).add_color

puts ColorStore.fetch_instance(Constants::BLUE)
ColorStore.fetch_instance(Constants::BLUE).add_color